### PR TITLE
Feat/timer/view model

### PR DESCRIPTION
**Pull request by francelu**
# Integrate Timer Feature

## Description
This PR introduces a new timer feature to the project, which includes the implementation of the timer model, data transfer objects, repository, view model, and associated tests. It integrates the logic for managing timers, including starting, stopping, and resetting timers, as well as synchronizing data with the Supabase backend.

It closes issue #217.  

### Note
This PR will not handle the delete functions on Timer VM since it's not useful for now. I'm creating a task in for later, when we will need it on the deletion of a user.  

I know I should have made a separate PR for the M from the VM. However, as I was implementing the VM, I had to change multiple times the M. That would have lead to a PR from M, and a PR for VM which would have overwritten the PR for M.  

This PR has as new lines a lot of documentation , and mainly tests code. Hence, the thousand lines. I'm sorryyy...

## Changes
- Added `Timer` data class with methods for serialization and state management.
- Introduced `TimerDto` data transfer object to facilitate conversion between the `Timer` model and API data (for Supabase).
- Defined `TimerRepository` interface and implemented `TimerRepositorySupabase` for API interactions.
- Added `TimerViewModel` for handling timer-related operations and UI state.
- Implemented `TimerManager` to manage timer actions (start, stop, reset) and have a persistent timer (running after app closed).

## Note
- The table in Supabase has been settled, with its policies. It only contains the list of last timers' elapsed times.
- I tried to fix the security hotspot raised by sonarcloud, no clue still...

#### Added
- `app/src/main/java/com/android/periodpals/model/timer/Timer.kt`
- `app/src/main/java/com/android/periodpals/model/timer/TimerDto.kt`
- `app/src/main/java/com/android/periodpals/model/timer/TimerManager.kt`
- `app/src/main/java/com/android/periodpals/model/timer/TimerModel.kt`
- `app/src/main/java/com/android/periodpals/model/timer/TimerModelSupabase.kt`
- `app/src/main/java/com/android/periodpals/model/timer/TimerViewModel.kt`
- `app/src/test/java/com/android/periodpals/model/timer/TimerDtoTest.kt`
- `app/src/test/java/com/android/periodpals/model/timer/TimerManagerTest.kt`
- `app/src/test/java/com/android/periodpals/model/timer/TimerModelSupabaseTest.kt`

#### Modified
- `app/src/main/java/com/android/periodpals/model/user/UserDto.kt`: I saw the documentation wasn't up-to-date when trying to understand who VM worked with Supabase.
- `app/build.gradle.kts`: add `viewBinding` in `buildFeatures`.

#### Removed
- None

## Dependencies Added
- None

## Testing
- Unit tests have been added for `TimerDto`, `TimerManager`, and `TimerRepositorySupabase` to ensure correct functionality.
- Tests verify serialization/deserialisation, timer actions, and API interactions with Supabase.

### CODE REVIEW

**Comment by francelu** [comment]
How the flow should be :
1. Timer UI: user taps on "Start", "Stop" or "Cancel" button.
2. Timer VM: receives start/stop/cancel event, then calls the corresponding method in `TimerManager`
3. `TimerManager`: starts / stops the `TimerService` and updates the `LiveData` (`timerState`)
4. `TimerService`:  runs in the background calculating elapsed time and periodically notifying `TimerManager` or `ViewModel`
5. UI update: VM updates UI via `LiveData`, s.t. timer display reflects timer state.

File: app/src/main/java/com/android/periodpals/model/timer/TimerModelSupabase.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      withContext(Dispatchers.Main) {
+        val result = supabaseClient.postgrest[TIMERS].delete { select() }.decodeSingle<TimerDto>()
+        Log.d(TAG, "deleteTimer: Success")
+        onSuccess(result)
+      }
+    } catch (e: Exception) {
+      Log.d(TAG, "deleteTimer: fail to delete timer: ${e.message}")
+      onFailure(e)
+    }
+  }
```
Comment (coaguila) [inline app/src/main/java/com/android/periodpals/model/timer/TimerModelSupabase.kt:64]: This is perhaps correct use of Supabase's `delete` function but perhaps a more correct version would come if we added an id to `Timer` & `TimerDto` and added a filter just like the [Supbase docs](https://supabase.com/docs/reference/kotlin/delete) recommend.

Something like this:
```suggestion
  override suspend fun deleteTimer(
      timerID: String, 
      onSuccess: (TimerDto) -> Unit, 
      onFailure: (Exception) -> Unit
  ) {
    try {
      withContext(Dispatchers.Main) {
        supabaseClient.postgrest[TIMERS].delete { 
          filter {
            eq("id", timerId)
          }
        }
        Log.d(TAG, "deleteTimer: Success")
        onSuccess(result)
      }
    } catch (e: Exception) {
      Log.d(TAG, "deleteTimer: fail to delete timer: ${e.message}")
      onFailure(e)
    }
  }
```

File: app/src/main/java/com/android/periodpals/model/timer/TimerModel.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  )
+
+  /**
+   * Deletes the timer for the specified user. RLS rules are applied to ensure that the user can
+   * only delete their own timer.
+   *
+   * @param onSuccess The callback to be invoked when the timer is deleted successfully.
+   * @param onFailure The callback to be invoked when an error occurs while deleting the timer.
+   */
+  suspend fun deleteTimer(onSuccess: (TimerDto) -> Unit, onFailure: (Exception) -> Unit)
```
Comment (coaguila) [inline app/src/main/java/com/android/periodpals/model/timer/TimerModel.kt:36]: Just adding to the suggestion in `TimerModelSupabase` : 
```suggestion
  suspend fun deleteTimer(timerID: String, onSuccess: (TimerDto) -> Unit, onFailure: (Exception) -> Unit)
```

File: app/src/test/java/com/android/periodpals/model/timer/TimerViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+//  fun loadTimerHasFailed() = runTest {
+//    doAnswer { it.getArgument<(Exception) -> Unit>(1)(Exception("failed")) }
+//        .`when`(timerModel)
+//        .loadTimer(any<(TimerDto) -> Unit>(), any<(Exception) -> Unit>())
+//
+//    timerViewModel.loadTimer()
+//
+//    assertNull(timerViewModel.timer.value)
+//  }
+// }
```
Comment (coaguila) [inline app/src/test/java/com/android/periodpals/model/timer/TimerViewModelTest.kt:186]: Plesase remove commented out code! If it is code you need for later you can get it later from git.

File: app/src/test/java/com/android/periodpals/model/timer/TimerViewModelTest.kt
Comment (coaguila) [inline app/src/test/java/com/android/periodpals/model/timer/TimerViewModelTest.kt:1]: Don't forget to add the roboelectric tags for View Model Test, don't hesitate to ask @charliemangano for help on how to add it to your tests.

File: app/src/test/java/com/android/periodpals/model/timer/TimerViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    val expected = timer.asTimer()
+
+    doAnswer { it.getArgument<(TimerDto) -> Unit>(0)(timer) }
+        .`when`(timerModel)
+        .loadTimer(any<(TimerDto) -> Unit>(), any<(Exception) -> Unit>())
+
+    timerViewModel.loadTimer()
+
+    assertEquals(expected, timerViewModel.timer.value)
+  }
```
Comment (coaguila) [inline app/src/test/java/com/android/periodpals/model/timer/TimerViewModelTest.kt:48]: Look into adding `verify` from Mockito, it allows us to test if the mocked function is called correctly with the correct parameters. Check some of the other tests.

File: app/src/main/java/com/android/periodpals/model/timer/TimerViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+  /** Retrieves the remaining time on the timer. */
+  fun getRemainingTime(): Long {
+    return timerManager.getRemainingTime()
+  }
+
+  @VisibleForTesting
+  fun setTimerForTesting(timer: Timer?) {
+    _timer.value = timer
+  }
```
Comment (coaguila) [inline app/src/main/java/com/android/periodpals/model/timer/TimerViewModel.kt:113]: So cool! I didn't know that there was `VisibleForTesting` annotation, this will be very useful in future development!

File: app/src/main/java/com/android/periodpals/model/timer/TimerModelSupabase.kt (hunk 1 above)
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerModelSupabase.kt:64]: Okay. I had thought about that. However, since in the RLS policies, a user is only allowed to insert, select, update, delete the timer linked to its uid, I thought we wouldn't need it. Sure I can do the changes ! :)

image

File: app/src/test/java/com/android/periodpals/model/timer/TimerViewModelTest.kt (hunk 3 above)
Comment (francelu) [reply app/src/test/java/com/android/periodpals/model/timer/TimerViewModelTest.kt:186]: Ohhhh my bad, I really just forgot to remove it ahaha xD  
Good you noticed !!

File: app/src/main/java/com/android/periodpals/model/timer/TimerViewModel.kt (hunk 5 above)
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerViewModel.kt:113]: I think Mistral AI helped me with that one, cause I was in so much despair :')

**Review by coaguila — CHANGES_REQUESTED** [review]
## Summary
Overall great changes but there are a couple problems. The quality of this code is superb and implementing the `@VisibleForTesting` annotation for testing. Also `TimeManager` functions return values instead of passing them through callbacks.
## Important (which includes)
#### Code Quality
`TimeManager` functions return values instead of passing them through callbacks. Also ` deleteTimer` in `TimerModelSupabase` could have a better implementation of Supabase's delete feature. Other than that, the code is well commented and very readable.
#### Functionality
Perhaps there should be a `getTimers` function for `TimerModel` & `TimerModelSupabase` that could fetch all alerts that could be better than storing all the timers in a string in a json format. Also `deleteTimer` in `TimerModelSupabase` should still take in an ID in case of problems with Supabase RLS. 
## Testing
Perhaps it would be useful to add a couple of `verify` to verify that the parameters are being passed on correctly.
## Documentation
LGTM
## Suggestions
This is great code needs improuvement, there's a couple of quality of life fixes such as not storing all the timers within the latest timer and also bits on commented code that no longer needs to be there.
## Steps before PR approved
- Add `timerID` to `Timer` & `TimerDto`
- Add `userID` to `Timer` & `TimerDto`
- Make `deleteTimer` in `TimerModel` & `TimerModelSupabase` check/filter for ID
- Create `getTimers` that fetches timers for`TimerModel`, `TimerModelSupabase` & `TimerViewModel`

File: app/src/test/java/com/android/periodpals/model/timer/TimerViewModelTest.kt
Comment (francelu) [reply app/src/test/java/com/android/periodpals/model/timer/TimerViewModelTest.kt:1]: Actually, the robolectric thing is just for the UI tests that we put in the unit tests. See this "Note" in [this](https://github.com/PeriodPals/periodpals/pull/208#issue-2668393254) PR description.

File: app/src/main/java/com/android/periodpals/model/timer/TimerModelSupabase.kt (hunk 1 above)
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerModelSupabase.kt]: Did the changes !

File: app/src/main/java/com/android/periodpals/model/timer/TimerModel.kt (hunk 2 above)
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerModel.kt]: Same here :)

**Review by coaguila — APPROVED** [review]
Thanks for implementing the changes, codes lgtm 👍🏽

**Review by charliemangano — CHANGES_REQUESTED** [review]
## Summary
Nice job with this! The implementation if very complete and well-structured.

## Important
#### Code Quality
The code is dense in functionality but still very readable. Nitpick but you could rename the `...fails` tests to `...Failure` for consistency.
#### Functionality
The code seems correct and passes all the tests, good job!
Take care to run the calls to Supabase within the main thread context in the model (see comments)
## Testing
Very complete testing! However, I don't understand the choice to use the `@VisibleForTesting` annotation instead of simply mocking a functionality, more on this in the comments.
Nitpick since your tests are already complete enough, but you could test for `null` values or extreme timer values, to cover more edge cases. 

## Documentation
Complete and up to date in most places, nicely done! See comments for places I noticed it was missing our outdated.

## Suggestions
See comments.

## Steps before PR approved
Please fix the model (see comments, in particular [this one](TODO)). See comments for nitpicks but I don't consider them necessary to implement before merging. Just make sure to open an issue for anything that you don't do right now.

File: app/src/main/java/com/android/periodpals/model/timer/TimerManager.kt [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+import java.util.Locale
+
+private const val TAG = "TimerManager"
+
+/**
+ * Manages the timer for the app.
+ *
+ * @param context The context used to access shared preferences.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerManager.kt:18]: Uhmm, I think I needed it when I was trying to make the timer stop automatically after 6 hours, for testing purposes. I'll remove it since the tests pass without.

File: app/build.gradle.kts [hunk 7]
```diff
@@ -31,7 +31,10 @@ android {
     vectorDrawables { useSupportLibrary = true }
   }
 
-  buildFeatures { buildConfig = true }
+  buildFeatures {
+    buildConfig = true
+    viewBinding = true
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/build.gradle.kts:36]: I though I had needed it, it was from the tutorial the coaches sent.  
I had removed it at one point, and some tests did not pass. But now I think we don't need it. All tests still pass without ! :))

File: app/src/main/java/com/android/periodpals/model/timer/TimerManager.kt [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+  private fun setTimerCounting(value: Boolean) {
+    timerCounting = value
+    with(sharedPref.edit()) {
+      putBoolean(COUNTING_KEY, value)
+      apply()
+    }
+  }
+
+  /**
+   * Starts the timer. The timer will run for a fixed duration of 6 hours.
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerManager.kt:90]: Updated !

File: app/src/main/java/com/android/periodpals/model/timer/TimerManager.kt [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+   *
+   * @param onSuccess The callback to be invoked when the timer is successfully started.
+   * @param onFailure The callback to be invoked when the timer fails to start.
+   */
+  fun startTimerAction(
+      onSuccess: () -> Unit = { Log.d(TAG, "startTimerAction: success callback") },
+      onFailure: (Exception) -> Unit = { e: Exception ->
+        Log.d(TAG, "startTimerAction: fail to start timer: ${e.message}")
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerManager.kt:98]: Changed, can you recheck ? :)

File: app/src/main/java/com/android/periodpals/model/timer/TimerManager.kt [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    const val PREFERENCES = "prefs"
+    const val START_TIME_KEY = "startKey"
+    const val STOP_TIME_KEY = "stopKey"
+    const val COUNTING_KEY = "countingKey"
+    const val COUNTDOWN_DURATION = 6 * 60 * 60 * 1000 // 6 hours in milliseconds
+  }
+
+  /** Starts the timer for testing purposes, failure case. */
+  @VisibleForTesting
+  fun startActionForTesting(onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerManager.kt:181]: Yeah really not ideal. I just couldn't find a way to mock failure, I'll retry...

File: app/src/main/java/com/android/periodpals/model/timer/TimerModel.kt [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  )
+
+  /**
+   * Deletes the timer with the specified ID.
+   *
+   * @param timerID The ID of the timer to delete.
+   * @param onSuccess The callback to be invoked when the timer is deleted successfully.
+   * @param onFailure The callback to be invoked when an error occurs while deleting the timer.
+   */
+  suspend fun deleteTimerByTimerId(
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerModel.kt:35]: I did the changes in Timer M and Timer M Supabase. However, I haven't yet implemented the VM delete functions

File: app/src/main/java/com/android/periodpals/model/timer/TimerModelSupabase.kt [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+ */
+class TimerRepositorySupabase(private val supabase: SupabaseClient) : TimerRepository {
+
+  override suspend fun addTimer(
+      timer: Timer,
+      onSuccess: () -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    try {
+      withContext(Dispatchers.IO) {
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerModelSupabase.kt:25]: I did the changes, can you check ? :)

File: app/src/main/java/com/android/periodpals/model/timer/TimerViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+private const val TAG = "TimerViewModel"
+
+/**
+ * View model for the timer feature.
+ *
+ * @property timerRepository The repository used for loading and saving timer data.
+ * @property timerManager The manager used for starting, stopping, and resetting the timer.
+ */
+class TimerViewModel(
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/model/timer/TimerViewModel.kt:16]: Updated !

File: app/src/test/java/com/android/periodpals/model/timer/TimerManagerTest.kt [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+import org.junit.Test
+import org.mockito.Mockito.anyBoolean
+import org.mockito.Mockito.anyString
+import org.mockito.Mockito.eq
+import org.mockito.Mockito.isNull
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.verify
+import org.mockito.Mockito.`when`
+
+class TimerManagerTest {
```
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/test/java/com/android/periodpals/model/timer/TimerManagerTest.kt:24]: Did the changes in the tests !

**Review by charliemangano — APPROVED** [review]
Thanks for implementing the changes, LGTM
